core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala

```diff
@@ -140,7 +143,16 @@ class JobProgressListener(conf: SparkConf) extends SparkListener with Logging {
     if (jobs.size > retainedJobs) {
       val toRemove = math.max(retainedJobs / 10, 1)
       jobs.take(toRemove).foreach { job =>
-        jobIdToData.remove(job.jobId)
+        for (
+          removedJob <- jobIdToData.remove(job.jobId);
+          jobGroupId = removedJob.jobGroup.orNull;
+          jobsInGroup <- jobGroupToJobIds.get(jobGroupId)
```

Comment:
this mix and match of `<-` and `=` took me a while to figure out what's going on. Can we rewrite this:
```
jobs.take(toRemove).foreach { job =>
  jobIdToData.remove(job.jobId).foreach { removedJob =>
    val jobGroupId = removedJob.jobGroup.orNull
    val jobsInGroup = jobGroupToJobIds.get(jobGroupId)
    jobsInGroup.remove(job.jobId)
    ...
  }
}
```